src/lib/utils/DY.Lib.SplitFunction.fst

```diff
@@ -0,0 +1,287 @@
+  // Correctness theorem on creating and applying a global function
+  apply_mk_global_fun: bare:(tagged_data_t -> output_t) -> x:tagged_data_t -> Lemma
+    (apply_global_fun (mk_global_fun bare) x == bare x);
+}
+
+/// Do a global function contain some given local function with some set of tags?
+/// This will be a crucial precondition for the correctness theorem `local_eq_global_lemma`.
+
+val has_local_fun: func:split_function_input_values -> func.global_fun_t -> (func.tag_set_t & func.local_fun_t) -> prop
+let has_local_fun func global_fun (tag_set, tagged_local_fun) =
+  forall tagged_data.
+    match func.decode_tagged_data tagged_data with
+    | Some (tag, raw_data) ->
+      tag `func.tag_belong_to` tag_set ==> (func.apply_global_fun global_fun tagged_data == func.apply_local_fun tagged_local_fun raw_data)
+    | None -> True
+
+/// Returns the first local function associated with a tag set containing `tag`, if it exists.
+/// In practice, only one tag set may contain `tag` because tag sets are mutually disjoint
+/// (c.f. precondition of `mk_global_fun_correct`).
+/// In that case, this function returns the *unique* local function associated with a tag set containing `tag`.
+val find_local_fun: func:split_function_input_values -> list (func.tag_set_t & func.local_fun_t) -> func.tag_t -> option func.local_fun_t
+let rec find_local_fun func tagged_local_funs tag =
+  match tagged_local_funs with
+  | [] -> None
+  | (h_tag_set, h_local_fun)::t_tagged_local_funs -> (
+    if tag `func.tag_belong_to` h_tag_set then
+      Some h_local_fun
+    else
+      find_local_fun func t_tagged_local_funs tag
+  )
+
+val mk_global_fun_aux: func:split_function_input_values -> list (func.tag_set_t & func.local_fun_t) -> func.tagged_data_t -> func.output_t
+let mk_global_fun_aux func tagged_local_funs tagged_data =
+  match func.decode_tagged_data tagged_data with
+  | Some (tag_set, raw_data) -> (
+    match find_local_fun func tagged_local_funs tag_set with
+    | Some tagged_local_fun -> func.apply_local_fun tagged_local_fun raw_data
+    | None -> func.apply_global_fun func.default_global_fun tagged_data
+  )
+  | None -> func.apply_global_fun func.default_global_fun tagged_data
+
+/// Given a list of tags and local functions, create the global function.
+
+[@@"opaque_to_smt"]
+val mk_global_fun: func:split_function_input_values -> list (func.tag_set_t & func.local_fun_t) -> func.global_fun_t
+let mk_global_fun func tagged_local_funs =
+  func.mk_global_fun (mk_global_fun_aux func tagged_local_funs)
+
+val memP_map:
+  #a:Type -> #b:Type ->
+  f:(a -> b) -> l:list a -> x:a ->
+  Lemma
+  (requires List.Tot.memP x l)
+  (ensures List.Tot.memP (f x) (List.Tot.map f l))
+let rec memP_map #a #b f l x =
+  match l with
+  | [] -> ()
+  | h::t ->
+    introduce x =!= h ==> List.Tot.memP (f x) (List.Tot.map f t)
+    with _. memP_map f t x
+
+val all_disjoint : #a:Type -> (a -> a -> prop) -> list a -> prop
```

Comment:
i think this function should have a more general name since it already has the general type.
maybe something like "for_allP_pairwise" ? (then it reads nicely when calling `for_allP_pairwise is_disjoint ...`)